### PR TITLE
Discord command testing

### LINKED ISSUE #60: Enable testing of Discord command coroutine functions
As part of User Story #51, set up a solution (mocks, second bot, whatever) to test the command functions that is compatible with code coverage as implemented in #53.

Completion criteria: Discord commands are testable and tests are written for at least one command and submitted to a pull request.

Due date: Monday, April 19

### PR DESCRIPTION
**Pull request by jcgoble3**
After a three-hour fight with asyncio, I present Discord command testing as ordered in Task #60. Do not expect to just run the test script and get results. In order to run this, you need:

- the `distest` package installed (using `pip`)
  - note that installing `distest` will **down**grade `discord.py` because `distest` apparently does not work with `discord.py` 1.7 yet
- a second bot user (requires creating another application in the developer portal and adding a bot user with admin permissions and privileged intents)
- a dedicated channel on your private server to run the tests on
- the `coverage` package installed using `pip` (only if you want to produce a test coverage report)

I don't expect everyone to set things up to run this. If you do want to run it, you will need to go into `utils/data.py` line 26 and change the long number to the user ID of the bot running the tests (not of the bot being tested). Then follow the instructions at the top of the test script.

If running it yourself, do not start the main bot separately; the test script will launch the tested bot itself in the same process as the test harness (this is critical as measuring test coverage cannot be done when the bot is running in a completely separate process, perhaps even on a different machine, from the test harness) and shut it down when the tests are complete. Even if you are not running coverage, you should still not start the bot separately because the script launches the bot itself regardless.

A few notes:

- Most of `index.py` was moved into `utils/bot.py`. This was necessary to make it possible to import the bot without automatically starting the infinite `bot.run` loop.
- `distest` is poorly documented, and so finding out how to do things is difficult. The best source for test assertions is the source code itself: https://github.com/JakeCover/distest/tree/develop/distest/TestInterface along with the sample tester in the root of that same repo.
- The changes in `utils/data.py` are required as the discord.py library by default ignores messages from other bots. We do not want the bot to respond to all bots, just the testing bot. The bot ID shown is my own testing bot; unfortunately there's no easy way to not have it hardcoded and I'm done fighting with this. If you want to run with your own testing bot, then as mentioned above, you will need to edit this (and make sure you revert the change. I think for the 9 days left in this semester, that will be okay.

### CODE REVIEW

File: distest-test-bot.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env python3.7
+
+# Based on the example tester at
+# https://github.com/JakeCover/distest/blob/develop/example_tester.py
+#
+# To run:
+# ./distest-test-bot.py --run all --channel <channel ID> 821891815329890336 <tester token>
```
Comment (jluckenbill) [inline distest-test-bot.py:7]: I like seeing the run command in the comment. Please consider adding this to the readme file.

Comment (jcgoble3) [reply distest-test-bot.py:7]: I plan to. See Action #78 (just created today) which I'll take care of around the end of next week as things are wrapping up.

File: utils/data.py [hunk 2]
```diff
@@ -13,11 +13,21 @@ def __init__(self, *args, prefix=None, **kwargs):
         self.trivia = Trivia()
 
     async def on_message(self, msg):
-        if not self.is_ready() or msg.author.bot or not permissions.can_handle(msg, "send_messages"):
+        if not self.is_ready() or not permissions.can_handle(msg, "send_messages"):
             return
 
         await self.process_commands(msg)
 
+    ## @story{51} In order to test command coroutines, a second bot is needed.
```
Comment (jluckenbill) [inline utils/data.py:21]: Why is a second bot needed?

**Review by jluckenbill — APPROVED** [review]
Code looks fine but could use more commenting.

Since there are some significant changes to the code base, namely relocation of sections of code, I think there needs to be more documenting of why these changes are necessary, rather than simply stating that they are necessary.

I'm approving this pull request because I don't see anything inherently wrong with the code and this is a class environment so I see no reason to deny or delay this pull request.

**Comment by sethsarchet** [comment]
Everything about your code looks really good and I see no problem merging it, but I happened to noticed an absence of Doxygen comments... Is there a reason for that, or am I just missing something?

File: utils/data.py (hunk 2 above)
Comment (jcgoble3) [reply utils/data.py:21]: A second bot is needed to issue commands to the tested bot in a controlled test server. This is black-box integration testing but relatively easy to set up. In order to do white-box testing or testing without a connection to Discord, a highly complex set of mocks would be needed, beginning with the context object but also literally everything its methods could possibly return. There have been a couple of attempts at this that can be found on GitHub, but they are woefully incomplete, contain serious bugs, and are barely alpha quality at best. Creating our own mocks was out of the question given the time constraints we have and the level of complexity involved, so the only reasonable option was distest, which requires a second bot.

**Comment by jcgoble3** [comment]
> Everything about your code looks really good and I see no problem merging it, but I happened to noticed an absence of Doxygen comments... Is there a reason for that, or am I just missing something?

@sethsarchet There's not much new code in the bot itself; the majority of changes to the bot itself is just cutting and pasting without modification to enable importing the bot. The new method override on the bot does have a Doxygen comment. The test code probably should have Doxygen comments, but I left them out because that code is going to change massively before I'm done.

**Comment by jcgoble3** [comment]
> Code looks fine but could use more commenting.

I know I'm bad at commenting.

> Since there are some significant changes to the code base, namely relocation of sections of code, I think there needs to be more documenting of why these changes are necessary, rather than simply stating that they are necessary.

Essentially the relocation is to enable the bot and config to be imported without starting the bot. Importing executes the entire file, so attempting to import index.py before would actually run the bot and never return. (I actually just thought of a cleaner way to do it, but it's not worth the code shuffle now.)

> I'm approving this pull request because I don't see anything inherently wrong with the code and this is a class environment so I see no reason to deny or delay this pull request.

Thanks.

**Pull request merged**